### PR TITLE
Add Resource.new(url: <url>) syntax back to yt

### PR DESCRIPTION
**Pull request by kangkyu**
I moved much code from url.rb to resource.rb to keep `url.resource` still work

### CODE REVIEW

File: lib/yt/models/resource.rb [hunk 1]
```diff
@@ -12,7 +12,13 @@ class Resource < Base
 
       # @!attribute [r] id
       #   @return [String] the ID that YouTube uses to identify each resource.
-      attr_reader :id
+      def id
+        if @url
```
Comment (dgb) [inline lib/yt/models/resource.rb]: Do we need this if/else here? In your code if you don't instantiate with a url, you've got `@id` set already.

File: spec/models/resource_spec.rb [hunk 2]
```diff
@@ -4,6 +4,13 @@
 describe Yt::Resource do
   subject(:resource) { Yt::Resource.new attrs }
 
+  context 'given a resource initialized with a URL (containing an ID)' do
+    let(:attrs) { {url: 'youtu.be/9bZkp7q19f0'} }
+
+    it { expect(resource.id).to eq '9bZkp7q19f0' }
```
Comment (dgb) [inline spec/models/resource_spec.rb:10]: Won't this make an HTTP request? If so, it seems like this test belongs in the `requests` dir.

File: lib/yt/models/resource.rb [hunk 3]
```diff
@@ -42,15 +48,20 @@ def unlisted?
...
+        end
         @auth = options[:auth]
         @snippet = Snippet.new(data: options[:snippet]) if options[:snippet]
         @status = Status.new(data: options[:status]) if options[:status]
       end
 
       # @private
       def kind
-        self.class.to_s.demodulize.underscore
+        (@match[:kind].to_s if @url) || self.class.to_s.demodulize.underscore
```
Comment (dgb) [inline lib/yt/models/resource.rb]: Can we change this to be an `if/else`? It's kind of hard to read as-is.

```ruby
if @url 
  @match[:kind].to_s
else
  self.class.to_s.demodulize.underscode
end

File: lib/yt/models/resource.rb [hunk 4]
```diff
@@ -66,8 +77,63 @@ def update(attributes = {})
...
+      # @return [Array<Regexp>] patterns matching URLs of YouTube channels.
+      CHANNEL_PATTERNS = [
+        %r{^(?:https?://)?(?:www\.)?youtube\.com/channel/(?<id>UC[a-zA-Z0-9_-]{22})},
+        %r{^(?:https?://)?(?:www\.)?youtube\.com/(?<format>c/|user/)?(?<name>[a-zA-Z0-9_-]+)}
+      ]
+
     private
 
+      def find_pattern_match
+        patterns.find(-> {{kind: :unknown}}) do |kind, regex|
```
Comment (dgb) [inline lib/yt/models/resource.rb]: I know this is stuff you moved around, but I find this kind of hard to follow.

File: spec/models/resource_spec.rb (hunk 2 above)
Comment (kangkyu) [reply spec/models/resource_spec.rb:10]: Does it? I think it doesn't because it's a video id

Comment (dgb) [reply spec/models/resource_spec.rb:10]: I might have just misunderstood how the url part works :)

Comment (kangkyu) [reply spec/models/resource_spec.rb:10]: By the way, we might need add more tests in `requests/` directory, for this change
